data/tutorials/lg_08_error_handling.md

```diff
@@ -125,67 +191,549 @@ let () =
+```ocaml
+match opt with
+| None -> ...    (* Something *)
+| Some x -> ...  (* Something else *)
+```
+However, sequencing such expressions leads to deep nesting which is often
+considered bad:
+
+> if you need more than 3 levels of indentation, you're screwed anyway, and
+> should fix your program.
+
+[Linux Kernel Style
+Guide](https://www.kernel.org/doc/Documentation/process/coding-style.rst)
+
+The recomended way to avoid that is to refrain from or delay attempting to
+access the content of an option value.
+
+### Using on `Option.map` and `Option.bind`
+
+Let's start with an example. Let's imagine one needs to write a function
+returning the [hostname](https://en.wikipedia.org/wiki/Hostname) part of an
+email address. For instance, given the email
+"gaston.lagaffe@courrier.dupuis.be," it would return "courrier."
+
+Here is a questionable but straightforward implementation using exceptions:
+```ocaml
+let host email =
+  let fqdn_pos = String.index email '@' + 1 in
+  let fqdn_len = String.length email - fqdn_pos in
+  let fqdn = String.sub email fqdn_pos fqdn_len in
+  try
+    let host_len = String.index fqdn '.' in
+    String.sub fqdn 0 host_len
+  with Not_found ->
+    if fqdn <> "" then fqdn else raise Not_found
+```
+This may fail by raising `Not_found` if the first the call to `String.index`
+does, which make sense since if there is no `@` character in the input string,
+signifying that it's not an email address. However, if the second call to `String.index` fails,
+meaning no dot character was found, we may return the whole fully qualified
+domain name (FQDN) as a fallback, but only if it isn't the empty string.
+
+Note that `String.sub` may throw `Invalid_argument`. Fortunately, this can't
+happen. In the worst case, the `@` character is the last one, then `fqdn_pos` is
+off range by one but `fqdn_len` is null, and that combination of parameters
+doesn't count as an invalid substring.
+
+Below is the equivalent function using the same logic, but using `Option` instead of
+exceptions:
+```ocaml
+let host_opt email =
+  match String.index_opt email '@' with
+  | Some at_pos -> begin
+      let fqdn_pos = at_pos + 1 in
+      let fqdn_len = String.length email - fqdn_pos in
+      let fqdn = String.sub email fqdn_pos fqdn_len in
+      match String.index_opt fqdn '.' with
+      | Some host_len -> Some (String.sub fqdn 0 host_len)
+      | None -> if fqdn <> "" then Some fqdn else None
+    end
+  | None -> None
+```
+
+Although it qualifies as safe, its legibility isn't improved. Some may even
+claim it is worse.
+
+Before showing how to improve this code, we need to explain how `Option.map` and
+`Option.bind` work.
+```ocaml
+val Option.map : ('a -> 'b) -> 'a option -> 'b option
+val Option.bind : 'a option -> ('a -> 'b option) -> 'b option
+```
+
+`Option.map` applies a function `f` to an option parameter, if it isn't `None`
+```ocaml
+let map f = function
+| Some x -> Some (f x)
+| None -> None
```

Comment:
I'll stop asking the same question to reduce the noise. Once I have clarity around these snippets of code that don't run or have outputs, whether they're not supposed to test it or if it belongs with another part, I'll revisit them. :)